src/ark/analysis/cell_neighborhood_stats.py

```diff
@@ -0,0 +1,227 @@
+import os
+import numpy as np
+import pandas as pd
+import xarray as xr
+from functools import reduce
+
+import ark.settings as settings
+from alpineer import misc_utils, io_utils
+
+
+def shannon_diversity(proportions):
+    """ Calculates the shannon diversity index for the provided proportions of a community
+    Args:
+        proportions (np.array):
+            the proportions of each individual group
+
+    Returns:
+        float:
+            the diversity of neighborhood
+    """
+
+    prop_index = proportions > 0
+    return -np.sum(proportions[prop_index] * np.log2(proportions[prop_index]))
+
+
+def compute_neighborhood_diversity(neighborhood_mat, cell_type_col):
+    """ Generates a diversity score for each cell using the neighborhood matrix
+    Args:
+        neighborhood_mat (pd.DataFrame):
+            the frequency neighbors matrix
+        cell_type_col (string):
+            the specific name of the cell type column the matrix represents
+
+    Returns:
+        pd.DataFrame:
+            contains the fov, label, cell_type, and diversity_cell_type values for each cell
+    """
+
+    misc_utils.verify_in_list(cell_type_column=cell_type_col,
+                              neighbor_matrix_columns=neighborhood_mat.columns)
+
+    # check input values
+    neighborhood_mat_values = np.array(neighborhood_mat.drop(
+            columns=[settings.FOV_ID, settings.CELL_LABEL, cell_type_col]))
+    if (neighborhood_mat_values > 1).any():
+        raise ValueError("Input must be frequency values.")
+
+    diversity_data = []
+    for fov in np.unique(neighborhood_mat[settings.FOV_ID]):
+        fov_neighborhoods = neighborhood_mat[neighborhood_mat[settings.FOV_ID] == fov]
+
+        diversity_scores = []
+        cells = fov_neighborhoods[settings.CELL_LABEL]
+        for label in cells:
+            # retrieve an array of only the neighbor frequencies for the cell
+            neighbor_freqs = \
+                fov_neighborhoods[fov_neighborhoods[settings.CELL_LABEL] == label].drop(
+                    columns=[settings.FOV_ID, settings.CELL_LABEL, cell_type_col]).values[0]
```

Comment:
We could probably swap the `for` loops and some of the indexing by using `groupby` and chaining it with `aggregate` methods and aggregating them. Up to you though!

Reply:
Would that be faster? This is actually from the other PR, I just moved it to a new file.

Reply:
It would be around a couple of orders of magnitude faster. I'm don't know if this step is a bottleneck, so speedups may not even matter, but my reasoning was mainly with respect to maintenance down the line and more readability by using more of the utilities that pandas provides.

Reply:
Okay, I just tested it and it's very quick as is, so Im going to leave it for now! I think there are a lot of places in the spatial analysis code that could be adjusted as you mentioned above, so we can circle back to that possibly in a clean up/speed up PR down the line.